### PR TITLE
[gcp][fix] Ensure proper handling of region and zone for all resources

### PR DESCRIPTION
**Pull request by anjafr**
# Description

# To-Dos

- [x] Add test coverage for new or updated functionality
- [x] Lint and test with `tox`
- [ ] Document new or updated functionality (someengineering/resoto.com#XXXX)

# Code of Conduct

By submitting this pull request, I agree to follow the [code of conduct](https://resoto.com/code-of-conduct).

### CODE REVIEW

File: plugins/gcp/resoto_plugin_gcp/resources/base.py [hunk 1]
```diff
@@ -168,29 +168,28 @@ def add_node(self, node: GcpResourceType, source: Optional[Json] = None) -> Opti
         log.debug(f"{self.name}: add node {node}")
         node._cloud = self.cloud
         node._account = self.project
-        if source is not None and InternalZoneProp in source:
-            if zone := self.zone_by_name.get(source[InternalZoneProp]):
-                node._zone = zone
-                node._region = self.region_by_zone_name[source[InternalZoneProp]]
-                self.add_edge(node, node=zone, reverse=True)
+        if not (node._region or isinstance(node, GcpRegion)):
```
Comment (aquamatthias) [inline plugins/gcp/resoto_plugin_gcp/resources/base.py:171]: - Why wouldn't we link a GcpRegion to a project? (or where does this happen)?
- If the resource already has a region, where is the edge created?

File: plugins/gcp/resoto_plugin_gcp/resources/compute.py [hunk 2]
```diff
@@ -3292,6 +3292,8 @@ def collect_individual(cls: Type[GcpResource], builder: GraphBuilder, zone: str,
             machineType=name,
         )
         machine_type_obj = GcpMachineType.from_api(result)
+        machine_type_obj._zone = builder.zone_by_name[zone]
+        machine_type_obj._region = builder.region_by_zone_name[zone]
```
Comment (aquamatthias) [inline plugins/gcp/resoto_plugin_gcp/resources/compute.py:3296]: So we collect machine types for a specific zone. To reflect the client behaviour for zonal responses, can't we simply do:
```python
result[InternalZoneProp]=zone
```
This should take care for setting the relevant properties and adding edges.

**Review by aquamatthias — APPROVED** [review]
Nice! One question if we should inline the function.

File: plugins/gcp/resoto_plugin_gcp/resources/base.py [hunk 3]
```diff
@@ -168,33 +168,53 @@ def add_node(self, node: GcpResourceType, source: Optional[Json] = None) -> Opti
...
-            # log.error(f"Neither zone nor region is set for node {source}, add to project.")
-            self.add_edge(node, node=self.project)
+
+        self.standard_edges(node, source)
+
         with self.graph_nodes_access:
             self.graph.add_node(node, source=source or {})
         return node
 
+    def standard_edges(self, node: GcpResourceType, source: Optional[Json] = None) -> None:
```
[the remark this answers is not in the document]
Comment (anjafr) [reply plugins/gcp/resoto_plugin_gcp/resources/base.py:178]: Its only use case is within `add_node`. I've moved the functionality into its own method mostly for testability reasons. There is a new test case that creates a blank `GcpProjectCollector`, injects a graph using some fixtures and then calls the new `standard_edges` method, without going through the full `collect` call.

Not sure how we'd keep that way of testing with the inlined function, but happy for any pointers / ideas!

File: plugins/gcp/resoto_plugin_gcp/resources/compute.py [hunk 4]
```diff
@@ -3291,6 +3291,7 @@ def collect_individual(cls: Type[GcpResource], builder: GraphBuilder, zone: str,
             zone=zone,
             machineType=name,
         )
+        result[InternalZoneProp] = zone
```
Comment (aquamatthias) [inline plugins/gcp/resoto_plugin_gcp/resources/compute.py:3294]: Please add a comment explaining this line of code. It is not obvious.

**Review by aquamatthias — APPROVED** [review]
thanks

File: plugins/gcp/resoto_plugin_gcp/resources/compute.py (hunk 4 above)
Comment (anjafr) [reply plugins/gcp/resoto_plugin_gcp/resources/compute.py:3294]: will do

**Pull request merged**